### ISSUE TITLE
Composer plugin changes are not applied to lockfile

### ISSUE THREAD

**Issue by gapple** [dependabot/dependabot-core#4030]
When a composer plugin modifies dependencies, those plugin changes are not accounted for in the `composer.lock` file generated by Dependabot.

In this particular case, [wikimedia/composer-merge-plugin](https://github.com/wikimedia/composer-merge-plugin) is used to merge dependencies from a sub-project into the root dependencies.
The PR created by dependabot doesn't appear to execute the plugin, resulting in packages being removed from the lock file.

**Package ecosystem**
composer
**Package manager version**
composer 2.1.3
**Language version**
PHP 7.4
**Manifest location and content prior to update**
https://github.com/gapple/dpbot-merge/blob/main/composer.json
https://github.com/gapple/dpbot-merge/blob/main/composer.lock
**dependabot.yml content**
https://github.com/gapple/dpbot-merge/blob/main/.github/dependabot.yml
**Updated dependency**
n/a
**What you expected to see, versus what you actually saw**
The `composer.lock` file changes in the dependabot PR removes the `gapple/structured-fields` library, which is merged from `/sub/composer.json`.
**Native package manager behavior**
Composer includes the `gapple/structured-fields` package information from `/sub/composer.json` in `/composer.lock`
**Images of the diff or a link to the PR, issue or logs**
https://github.com/gapple/dpbot-merge/pull/1
**🕹 Bonus points: Smallest manifest that reproduces the issue**

I created a minimal repository to replicate the issue: https://github.com/gapple/dpbot-merge

**Comment by gapple** [comment]
This issue originated in a private repository, where for some reason dependencies were moved from `packages` to `packages-dev` in composer.lock, rather than removed, but it appears to be from the same cause (composer-merge-plugin not being executed)

**Comment by feelepxyz** [comment]
> dependabot doesn't appear to execute the plugin

This sounds very likely. We've re-implemented a bunch of composer update logic which probably doesn't set up plugins in the right way. The code is here in case you have some spare cycles to help out: https://github.com/dependabot/dependabot-core/blob/main/composer/helpers/v2/src/Updater.php

**Comment by kubawerlos** [comment]
Quick investigation show that [this](https://github.com/composer/composer/blob/2.1.4/src/Composer/Installer.php#L339) is the only place where `POST_UPDATE_CMD` event is dispatched and as we can see it's in `if ($this->runScripts)` condition.

So it looks that it's not possible to run plugins when running scripts is disabled, yet it is possible - I've tested it with Composer on CLI and it works fine.

@Seldaek can you hint us what is wrong here (here = the link from the above comment), please?

**Comment by Seldaek** [comment]
runScripts in Installer is never disabled anymore as of Composer 2.1.2, it's just there for BC for people integrating with the code, and for dry-run support. 

Anyway I suspect this line should be deleted https://github.com/dependabot/dependabot-core/blob/main/composer/helpers/v2/src/Updater.php#L86 but I am not familiar enough with dependabot to know the complete implications.

**Comment by T2L** [comment]
Affected by this as well. Afaik this happens because dependabot does not run scripts => thus Composer does not dispatch events => most plugins act on (missing) events.

@feelepxyz what's the reason behind disabling plugins ?

**Comment by jurre** [comment]
> what's the reason behind disabling scripts ?

It's two-fold:
- minimizing the work that dependabot needs to do in order to perform an update (composer updates are already very slow)
- preventing potentially malicious code from executing in the dependabot environment and exfiltrating source code to bad actors

**Comment by T2L** [comment]
I see your point and can add that

- performance of composer 2 is order of magnitude better, and normally scripts are super fast and don't cause any CPU usage at all
- this is a valid point, not sure what can be done here, sandboxing and/or limiting running time
- scripts in 99% work with other files in the repo, so if it's not being checked out -> scripts usually cannot do anything

**Comment by jurre** [comment]
Yeah we've seen the performance improvements and they're great!

We do sandbox jobs, so our infrastructure and any secrets should be well protected, but source code exfiltration is still a concern. We do use tokens with the least possible privileges and they expire after a certain amount of time, but it's still a potential risk that I would like to understand before committing to it.

I am not super familiar with the php ecosystem (but have worked on our composer integration a bit), could you maybe point me to some examples of popular libraries that use such scripts? It'll help me understand the usecase.

For example, I wonder if a GitHub Action could be run on Dependabot branches and commit the results back to the PR branch. If the commit starts with `[dependabot skip]`, dependabot will still be able to rebase the PR etc.

**Comment by T2L** [comment]
So, technically there are 2 different things:

- Composer scripts => those could be anything, including system calls (like `cp`)  => so clearly there is an issue. To be honest usually scripts are not *that* important during updates, as the usual things is like to create some symlinks, remove example code (that is being shipped sometimes with a package etc)
- Composer plugins => those are composer packages themselves => they can respond to different Composer events and act upon them. Some popular plugins are: https://github.com/wikimedia/composer-merge-plugin, https://github.com/cweagans/composer-patches, https://github.com/composer/installers. 


Technically, any of those things could run malicious code I'm afraid, but to be honest existing implementation does not forbid plugins, but there was an issue in Composer preventing running them when scripts are disabled ¯\_(ツ)_/¯

Back to the issue, the problematic plugin is https://github.com/wikimedia/composer-merge-plugin which recalculates dependencies in realtime, and dependabot does not run it, so instead of updating a single dependency - it removes a lot of things

**Comment by T2L** [comment]
btw, I see https://docs.github.com/en/code-security/supply-chain-security/keeping-your-dependencies-updated-automatically/configuration-options-for-dependency-updates#insecure-external-code-execution

seems like the same/very similar thing we have here, no ?

**Comment by jurre** [comment]
> seems like the same/very similar thing we have here, no ?

Yeah I think that's definitely similar, but it's also been a point of confusion for users, so if we can avoid having to add that that'd be good IMO.

**Comment by gapple** [comment]
Composer 2.2 introduced an `allow-plugins` config value that would prevent new/unknown plugins from executing
https://blog.packagist.com/composer-2-2/#more-secure-plugin-execution

**Comment by gapple** [comment]
> For example, I wonder if a GitHub Action could be run on Dependabot branches and commit the results back to the PR branch. If the commit starts with [dependabot skip], dependabot will still be able to rebase the PR etc.

Because packages get removed from composer.lock, just running `$ composer update --lock` after Dependabot's commit can result in any of the removed & re-added dependencies getting updated as well, when they wouldn't be by a manual update.  To limit additional updates, as far as I can tell, an action would have to rollback Dependabot's update and re-executed the same update for the target package (maybe parsing the package name from dependabot's commit message?).

**Comment by gapple** [comment]
> For example, I wonder if a GitHub Action could be run on Dependabot branches and commit the results back to the PR branch. If the commit starts with [dependabot skip], dependabot will still be able to rebase the PR etc.

I've implemented this on my test repository, and it works pretty well:
https://github.com/gapple/dpbot-merge/blob/main/.github/workflows/dependabot-composer.yml

One catch is workflows won't be triggered on the fix commit with the default GITHUB_TOKEN, so I needed to setup a separate Personal Access Token - it would be nice if I could grant an additional permission to the GITHUB_TOKEN instead since there is sufficient conditions to prevent recursive workflow runs.

**Comment by driskell** [comment]
I raised this on the composer merge plug-in side as I wondered if this is more a problem with the way it merges. Like maybe it should be committing the merge into the parent composer JSON. TBF though that’s not going to work as it could be merging in from a package and if it gets updated…

The GitHub action approach is best. I already use it because there’s far more than merge plugin to consider. For example drupal composer scaffold plugin sets up drupal files and then you have symfony flex recipes setting up config files. So it’s always good to have an action that either updates dependabot PRs or (in my case) raises another PR after merge to repair the composer lock and commit any scaffold changes.

It’s just the moment you decide to support merge plugin you then need:

- Repo must be checked out to get at the merged files potentially

- Plugins and recipes can change other files, for example drupal scaffold plugin and symfony flex recipes would create new config files and these would need adding to the commit also

- Not only would plug-ins need to run but as I remember dependabot speeds up the run by skipping downloading packages since it doesn’t need to be installing it just needs the dependency solution, so now you need to download 100s of MBs of packages for a Drupal project as well as checkout repo, because merge plugin might reference a composer file from another installed package

I guess you could make it “opt-in” and tell dependabot which packages it should install to keep things in sync but it seems a bit tedious and risks hitting timeouts whereas using an Action can take as long as it needs.

**Comment by jeffwidman** [comment]
I think @driskell raises some really good points about the difficulties here for native Dependabot vs running it as a custom GitHub action. In particular, I'd be concerned about the downloads resulting in a lot more timeouts, which are [already a problem for Drupal](https://github.com/dependabot/dependabot-core/issues/2999).

Given that the workaround of a custom action exists, and @gapple nicely provided example code in https://github.com/dependabot/dependabot-core/issues/4030#issuecomment-1079818337, I'm tempted to close this as "a known problem, but one that we see no realistic way to fix".

That said, I'm open to input from others involved with this thread... what say you?

**Comment by jeffwidman** [comment]
On second thought, we could/should make this configurable so that folks running `dependabot-core` standalone (ie, outside of native GitHub Dependabot) could set an env var or a config param to enable the `allow-plugins` option.

It will likely stay a "wontfix" for GitHub native for the forseeable future due to the reasons I mentioned above.

**Comment by kubawerlos** [comment]
@jeffwidman would love that, my current workflow is whenever Dependabot raises PR, I bump dependencies in separate PR and see Dependabot one closed.

Being able to simply click "merge" would be a great improvement.

**Comment by jeffwidman** [comment]
Want to submit a PR? This isn't something the core team can justify spending much time on, but we're always open to community contributions.
